sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/ExponentialCounterFactory.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.internal.state;
+
+public interface ExponentialCounterFactory {
```

Comment:
The DataDog histogram swaps between Dense/Sparse counter implementations.   However, they're using double + partial counts for shuffling/resizing dynmaically which isn't an option for us.

I left the ability for us to mix/match dense (circular buffer) with sparse (map counter) to see if we can get any gains, but I wasn't able to find a scenario where this was performant, especially given how we expose dense-buckets in OTLP, and our most likely use case (latency monitoring) tends to follow normal-distributions.

Reply:
Also, I started wiring this through, and it actually needs the type to embedd pretty far up into the hierarchy to the point you have `ExponentialHistogramAggregation<T>`.   where T = the counter type.   It starts to feel pretty clunky and wierd.  Let me know if:

1. You'd like me to continue with that change.
2. I can just delete MapCounter and remove a lot of the factory/strategy options.
3. We leave it as it is.

Reply:
We can remove MapCounter in this PR if it makes things easier

Reply:
Don't feel strongly about this. Is it useful to keep `MapCounter` around for future performance comparisons or has it officially run its course? 

Can also just have `public AdaptingCircularBufferCounter(ExponentialCounter toCopy)` throw if `!(toCopy instanceOf AdaptingCircularBufferCounter)`.

Reply:
Sorry for AFK on this.  You'll have to bear with my availability this month :(.    

My opinion is that `MapCounter` is sufficiently poorly performing I think we can oust it for now.   There might be a world where spare-vector could perform well here, but not with the currently expected defaults of exponential histograms (specifically 320 buckets).   Today though, it puts serious strain on the garbage collector as we're duplicating a large map of atomic integers frequently.

Given @anuraaga's general concerns, I've still been spending time thinking through alternative implementations and mechanisms here.   Specifically, I want to prioritize a few optimisations in metrics:

1. Figuring out a way to avoid as many allocations in the collection path.   As @anuraaga called out, allocating 5MB of RAM every 1 minute (or less) is unacceptable.
2. Trying to leverage "dumb java code" that hotspots well as much as possible.  While I still think the layered integer is necessary for RAM savings at our current bucket defaults, I would still love a simpler implementation if it didn't come with 2-4x overhead in RAM.

I'll try to shore up this PR's remaining comments.   Right now, it looks like I should be able to attend the SiG this week, so perhaps we can discuss in person future work.

Reply:
Actually, I had some ideas where having `MapCounter` around (for the collection path) may be a decent optimisation.  Specifically for low-frequency histogram timeseries, if we detect # of buckets with values < 5 e.g. (or count is 0) we can optimise the collection path by NOT allocating a new huge integer array and just returning the Sparse (or empty) counter implementation.  However, I'll want to benchmark that to make sure it's actually worth it.

I'd like to shore up  other comments in this PR (for toCopy, it's actually important it works on all ExponentialCounter implementations), and then I'll work on a new PR that measures performance during collection we can use to identify and tune allocations in that path.